adrs/0541531435/README.md

```diff
@@ -0,0 +1,118 @@
+contributing to projects, and sometimes relying on a particular group of people
+that know the unspoken/undocumented rule.
+
+In other completely valid cases, like putting everything at the root of the
+project, don't take into consideration caveats like having so many metadata such
+as `.gitignore`, `.prettierrc`, `.eslintrc`, `.markdownlint.json`,
+`tsconfig.json` files and many other files and directories.
+
+Subjectively speaking, they add noise and hard to navigate or concentrate when
+put all your files at the root of the project.
+
+Furthermore,
+
+* Where do I suppose to put my test files? What are the differences between X and
+Y directories?
+
+* How to structure the project when I have multiple testing frameworks, or
+multiple documentation tools?
+
+This ADR is an attempt to follow the steps of other ecosystems like Elixir,
+Ruby or Rust came to a consensus about this topic.
+
+### About compilations
+
+We should tell what requires compilation in the project, which is extremely
+useful in library projects since you should be aware of such detail to consume
+the project. Or in cases that you would like to run the project in production.
+
+Having a directory that reflects such information is crucial to mitigate mistakes
+in application deployments, or misuse of libraries.
+
+### About test frameworks
+
+Sometimes you may have multiple test frameworks for your project, most likely
+for `end-to-end`, `integration`, or `unit` tests, but not limited to those
+concepts. Therefore, people have tried to come up with `./tests/integration`,
+`./tests/unit`, `./tests/end-to-end`.
+
+Although that may be a clean way to structure your tests, the problems arrive
+when you try to articulate the differences between those categories without
+getting into an in-depth conversation about X vs. Y and end up with unclear
+boundaries, relying too much on the programmer's experiences to articulate the
+most optimal definition or decision.
+
+Additionally, if you try to migrate your project to another testing tool that
+falls into one of those possible categories, you will mingle tests that are
+run differently depending on the underline tool, and that may be really
+important in some cases.
+
+### About documentation
+
+Documentation tools are amazingly diverse in the ecosystem; such tools require
+different configurations and code structure, therefore mingle tools may cause
+incompatible situations, or confusing documentation source code.
+
+Also, it is unclear today what should I do if I have multiple documentation
+websites (internal or public as an example), and how to manage them.
+
+## Resolution
+
+* You **must** follow the following directory structures:
+
+  ```text
+  .
+  ├── package.json
+  ├── bin
+  ├── dist
+  ├── docs
+  │   ├── storybook
+  │   └── website
+  │   └── website-next
+  │   └── [another website or tool]
+  ├── examples
+  ├── lib
+  ├── src
+  └── tests
+      ├── cypress
+      └── jest
```

Comment:
One thing came out in our chat yesterday: you could do full-stack testing in `./tests/jest`.   

Some care should be given to separating fast-running unit & integration tests from slow-running full-stack tests.  Full-stack tests are higher on the test pyramid; ideally, we want to run expensive tests less frequently and asynchronously -- like on CI.  

Recommending a specific strategy is out-of-scope of this ADR.  In passing, I'll mention a few options to achieve this:  1) use an e2e directory `./tests/jest/e2e/`, or 2) tag your tests and add a script to allow/block target tags.